src/comments.ts

```diff
@@ -0,0 +1,297 @@
+        repositoryGitCloneURL: remoteUrl,
+        path,
+        branch,
+        selection,
+    }
+    const input: SourcegraphGQL.IDiscussionThreadCreateInput = {
+        title: text,
+        contents: text,
+        targetRepo,
+    }
+
+    const data = await mutateGraphQL(
+        gql`
+            mutation CreateThread($input: DiscussionThreadCreateInput!, $relativeRev: String!) {
+                discussions {
+                    createThread(input: $input) {
+                        ...DiscussionThreadFields
+                    }
+                }
+            }
+            ${discussionThreadFieldsFragment}
+        `,
+        { input, relativeRev: branch }
+    )
+
+    if (!data.discussions || !data.discussions.createThread) {
+        throw new Error(`Invalid GraphQL response for CreateThread`)
+    }
+
+    return discussionToCommentThread(document, range, data.discussions.createThread)
+}
+
+async function replyToCommentThread(
+    document: vscode.TextDocument,
+    range: vscode.Range,
+    thread: vscode.CommentThread,
+    text: string
+): Promise<vscode.CommentThread> {
+    const [, branch] = await repoInfo(document.fileName)
+    const data = await mutateGraphQL(
+        gql`
+            mutation AddCommentToThread($threadID: ID!, $contents: String!, $relativeRev: String!) {
+                discussions {
+                    addCommentToThread(threadID: $threadID, contents: $contents) {
+                        ...DiscussionThreadFields
+                    }
+                }
+            }
+            ${discussionThreadFieldsFragment}
+        `,
+        { threadID: thread.threadId, contents: text, relativeRev: branch }
+    )
+
+    if (!data.discussions || !data.discussions.addCommentToThread) {
+        throw new Error(`Invalid GraphQL response for AddCommentToThread`)
+    }
+
+    return discussionToCommentThread(document, range, data.discussions.addCommentToThread)
+}
+function discussionToCommentThread(
+    document: vscode.TextDocument,
+    range: vscode.Range,
+    thread: SourcegraphGQL.IDiscussionThread
+): vscode.CommentThread {
+    const comments: vscode.Comment[] = []
+    for (const comment of thread.comments.nodes) {
+        comments.push({
+            commentId: comment.id,
+            body: new vscode.MarkdownString(comment.contents),
+            userName: comment.author.displayName || comment.author.username, // only username?
+            gravatar: comment.author.avatarURL || '',
+        })
+    }
+    return {
+        threadId: thread.id,
+        resource: document.uri,
+        range,
+        comments,
+    }
+}
+
+function getSelection(
+    document: vscode.TextDocument,
+    range: vscode.Range
+): SourcegraphGQL.IDiscussionThreadTargetRepoSelectionInput {
+    const beforeRange = new vscode.Range(range.start.line - 3, 0, range.start.line - 1, 0)
+    const linesBefore = getLines(document, beforeRange)
+
+    const lines = getLines(document, range)
```

Comment:
I believe the capturing of context lines here is wrong. This is one of the issues I saw in your PR before but I didn't have time to investigate it.

Additionally, before we start using this against dogfood/prod I need to figure out whether or not you should actually be specifying potentially modified context lines here AND specifying an exact Git revision, or if you should be doing something else in that situation. Otherwise, this will leave our data in an inconsistent state and it will be hard to deal with.

Reply:
> I believe the capturing of context lines here is wrong.

Do you mean the logic is wrong, or do you mean we shouldn't be capturing them at all?

VS Code does have a case that web doesn't have to deal with: unpushed (or even unsaved) modifications. If we only want to support discussing committed/pushed code (which is fine), then there is some work to be done to limit the commenting ranges (as discussed in another comment).

Reply:
When I last tested, the logic was wrong in some way (but I forget exactly how. I will check / reconfirm this logic is solid and that we're capturing what I expect here. I'll comment back here once I've done this.

I've been thinking about the issue of unpushed/unsaved modifications, and I see a few options (in no particular order):

1. Do not allow commenting on unpushed/unsaved modifications. I assume this would be hard to implement. It would also be unfortunate / not a good long term solution as it would prevent e.g. asking questions about code you are actively developing (something we could immediately support with e.g. email notifications).
2. Allow commenting on unpushed/unsaved file modifications, but clearly denote it by e.g. passing two new fields to the backend when we create threads indicating "this is on code that is modified in the editor" and "this is code that is committed to Git, but may not be pushed to the remote". This lets us e.g. warn the user in the web UI. This seems like the most ideal outcome, but I am unsure how hard detecting these two cases would be. For the first flag, we would need to ask VS Code if the file is modified in their editor somehow (even just a single "is modified" boolean would do, we don't need to do individual line ranges). For the second flag, we can detect this on the backend when creating a comment with a Git revision that doesn't exist in the repository to our knowledge yet.
3. Don't do anything here yet, and just pass a flag to the backend when creating comments indicating "this comment came from VS Code". This would let us defer the decision until later while not e.g. harming the web experience.

Of the above I think option 2 would be the best, and probably relatively simple to implement.

Reply:
> Of the above I think option 2 would be the best, and probably relatively simple to implement.

I agree and I wrote similar logic before in the old Sourcegraph Editor comments implementation. I think it reduces to sending the newest commit hash that the selected lines blame to. If this includes uncommitted lines, then the hash is either omitted or uses a sentinel value `00000000` like what git does. If some lines are committed but not pushed, then the web can detect that (e.g. revision won't be found). This gracefully handles comments on commits that aren't yet pushed but will be.

Reply:
Using the sentinel value `00000000` SGTM, but we should document this in the GraphQL API since we will only accept one literal string value (e.g. not `000000000000000000000000000000000000000000` or a different number of zeros like Git does). I will document this. Are you aware of any online Git documentation about this value that I can link to? I couldn't find any.

> If some lines are committed but not pushed, then the web can detect that (e.g. revision won't be found).

I agree, this is the best approach. It also means zero work here for that case.

TODO:

- [ ] @slimsag to update the GraphQL API docs to state the sentinel value `00000000` meaning.
- [ ] @nicksnyder to add the logic in this PR to send `00000000` when comment is on uncomitted code, and send correct commit (or `00000000` "for now" as a hack) when commenting on committed but unpushed code.

To merge, I am fine with something as simple as "If the file is at all modified, we send commit `00000000`" but as the web UX improves it would suck if we have lots of discussions created this way, so we should fix this relatively soon / before we start using this a ton.

Reply:
> Are you aware of any online Git documentation about this value that I can link to?

No, but it is basically the commit hash of only zeros so forty zeros is the "canonical" value (everything else is just an abbreviation like you can do with any commit hash).

Do you prefer this sentinel value over just not sending a value at all (the later might be more intuitive from an API perspective)?

Reply:
Jeez, thank you for questioning me on that. This is exactly why I originally made the field nullable, so yeah, we should just send null in that case:

https://github.com/sourcegraph/sourcegraph/blob/1d7722a5295141024f3d2b0e32be790a04827112/cmd/frontend/graphqlbackend/schema.graphql#L387-L389

Reply:
I came here to ask too if we can just use null. Sentinel values are bad for type safety.